### PR TITLE
builtins: Extend MemRef to cover more of the MLIR dialect

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR aims to get better memref support into xDSL

I want to implement with this PR:

 - `memref.dim`
 - `memref.rank`

Not added in this PR, but might be of interest in the future.
 - `memref.view`and `memref.subview`
 - `memref.reshape`
 - `memref.realloc`

### CODE REVIEW

File: tests/dialects/test_memref.py
Comment (georgebisbas) [inline tests/dialects/test_memref.py:123]: maybe name dim0 or sth similar....

File: tests/dialects/test_memref.py
Comment (superlopuh) [inline tests/dialects/test_memref.py:163]: why disable yapf?

File: xdsl/dialects/memref.py
Comment (superlopuh) [inline xdsl/dialects/memref.py:186]: How about this?

```suggestion
            shape: Optional[List[int | AnyIntegerAttr]] = [1],
```

And later `shape = list(shape)`. Making it Optional seems a bit ugly, and if we're good about making copies then it seems like the list copy is not a big penalty.

File: tests/dialects/test_memref.py
Comment (superlopuh) [inline tests/dialects/test_memref.py:169]: This feels funky, what's the test exactly? Feels like we might be better off doing small unit tests that don't involve printing or parsing

File: tests/dialects/test_memref.py
Comment (AntonLydike) [reply tests/dialects/test_memref.py:163]: Because it makes it unreadable otherwise :/

File: xdsl/dialects/memref.py
Comment (AntonLydike) [reply xdsl/dialects/memref.py:186]: I don't really like this. I think this is too easy to break, by removing an "unnecessary" `list(...)` cast... We should stick with the default of `None + if`.

File: tests/dialects/test_memref.py
Comment (AntonLydike) [reply tests/dialects/test_memref.py:169]: But this is done in lots of other places as well. Maybe we should convert this to a filecheck?

Comment (AntonLydike) [reply tests/dialects/test_memref.py]: I added a file check for the output, and a verify check to run all the verifiers over it, if that's okay. I don't know if we should have this as a test, but @georgebisbas requested to have this matmul example added as a test.

We might want to move it to a notebook in the future.

**Review by georgebisbas — APPROVED** [review]

File: xdsl/dialects/memref.py
Comment (PapyChacal) [inline xdsl/dialects/memref.py:290]:
```suggestion
                              index: SSAValue | Operation):
```

**Review by PapyChacal — CHANGES_REQUESTED** [review]
Requesting minor changes. Should we add MLIR Interop tests?

File: xdsl/dialects/memref.py
Comment (AntonLydike) [reply xdsl/dialects/memref.py:187]: The problem is this:
image

Here the second `test` function uses the fix @superlopuh proposed. Now if you look at the function body, we see a `list` cast, that does "nothing" (in fact `assert output == input` will hold!). But if this supposed nop is removed in a future commit, it might produce errors that are difficult to debug.

The `None + if` approach is a widely accepted practice in pyhton and we should stick with that imo.

File: tests/dialects/test_memref.py
Comment (webmiche) [inline tests/dialects/test_memref.py:141]: I personally disagree with having this as our tests. I feel it does not convey actual useful information to hunt bugs with but is only a big use case. Maybe I am missing something though.

File: tests/filecheck/dialects/memref/matmul.mlir
Comment (webmiche) [inline tests/filecheck/dialects/memref/matmul.mlir:2]: Same issue here.

**Review by webmiche — CHANGES_REQUESTED** [review]
Yes, please add interoperation tests to check that we can actually interface with MLIR.

File: tests/dialects/test_memref.py
Comment (math-fehr) [reply tests/dialects/test_memref.py:141]: I personally think that we should make a pass on all of our tests, to clean these tests, and move them to "integration tests" or examples.
But right now, in our repo, we do have tests like this.
I'm quite busy in the next week, but maybe after I can think of the test dialect?

In the meantime, I would say that it is fine to have this here for now, since we need to clean up the entire repo anyway.
Though maybe you feel strongly about the contrary?

File: xdsl/dialects/memref.py
Comment (math-fehr) [reply xdsl/dialects/memref.py:187]: I'm fine for both options, but:
* If we take the `list(...)` approach, we should add a `#copy` comment
* If we take the `= None` approach, we should probably document the default value in the function.

File: xdsl/dialects/memref.py
Comment (math-fehr) [inline xdsl/dialects/memref.py:283]: Shouldn't this be `MemRefType[Attribute]` instead?
`MemRefType` is a generic type.

File: xdsl/dialects/memref.py
Comment (math-fehr) [inline xdsl/dialects/memref.py:298]: Same here with `MemRefType`

**Review by math-fehr — COMMENTED** [review]
Nice! I added some comments, but this is otherwise good!

File: xdsl/dialects/memref.py
Comment (AntonLydike) [reply xdsl/dialects/memref.py:187]: I am wondering why we even have a default value of `[1]` here. The mlir one doesn't have a default size afaik: https://mlir.llvm.org/docs/Dialects/MemRef/#memrefalloca-mlirmemrefallocaop

**Comment by PapyChacal** [comment]
> So, the interop tests fail because we don't have `fastmath`! What's the best way around that?

Adding it :p
I would be happy with a mock implementation to only parse and handle the default `#arith.fastmath<none>`, no need to handle all its uses directly.

**Comment by webmiche** [comment]
I guess this therefore now depends on #458, right? I would prefer to wait until we merged fastmath flags before merging this.

**Comment by AntonLydike** [comment]
So there's #458 open to add just that, but it's an `enum` thingy, which the parser is currently struggling with... This might not be super trivial.

File: tests/dialects/test_memref.py
Comment (webmiche) [reply tests/dialects/test_memref.py:141]: Merging as is is okay with me, though I am not sure when we will get to actually fixing this (like the other about 2000 issues we have discussed before :wink: )

**Comment by AntonLydike** [comment]
I removed the `arith` bits for now so we can have this merged, I'll ad it once we merge the `fastmath` opts

**Comment by webmiche** [comment]
Hmm, true. I can't quite estimate the effort needed to get this to work...

Considering this is not actually an issue on the ops added here, but rather the already existing ops in arith, couldn't you write an interoperation test that does not need `mulf`? Maybe redefine the matrix multiply on an integer matrix, if you want to keep it.

(side note: introducing dependecies in the tests can make life hard sometimes, can't it :wink: ).

**Comment by PapyChacal** [comment]
The Enum situation is not blocking here, `#arith.fastmath<none>` can be parsed as a ParameterizedAttribute. Just wanted to clarify, but I don't have a strong opinion on whether you should wait for the other PR or not.

**Comment by AntonLydike** [comment]
> The Enum situation is not blocking here, `#arith.fastmath<none>` can be parsed as a ParameterizedAttribute. Just wanted to clarify, but I don't have a strong opinion on whether you should wait for the other PR or not.

I really don't want to introduce merge conflicts for that PR for something that's completely optional in this one. I moved over to i64 because that way we dodge all of this ^^

**Review by webmiche — APPROVED** [review]

**Comment by AntonLydike** [comment]
@PapyChacal It says I can't merge because you requested changes, but I can't see them and can't request a re-review from you. Can you see if you can press anything?

**Review by PapyChacal — APPROVED** [review]

**Pull request merged**
